tools/create_service_bundle.py

```diff
@@ -0,0 +1,568 @@
+
+
+def browse_executor_tasks(agent_id: str, executor_sandbox_path: str) -> List[dict]:
+    executor_sandbox = browse_executor_sandbox(agent_id, executor_sandbox_path)
+    tasks_directory = next(
+        filter(
+            lambda f: f["mode"].startswith("d") and f["path"].endswith("/tasks"),
+            executor_sandbox,
+        ),
+        None,
+    )
+
+    if tasks_directory:
+        return browse_agent_path(agent_id, tasks_directory["path"])
+    else:
+        return []
+
+
+@retrying.retry(
+    wait_fixed=DEFAULT_RETRY_WAIT, stop_max_attempt_number=DEFAULT_RETRY_MAX_ATTEMPTS
+)
+def browse_task_sandbox(
+    agent_id: str, executor_sandbox_path: str, task_id: str
+) -> List[dict]:
+    executor_tasks = browse_executor_tasks(agent_id, executor_sandbox_path)
+
+    if executor_tasks:
+        task_sandbox_path = os.path.join(
+            executor_sandbox_path, "tasks/{}/".format(task_id)
+        )
+        return browse_agent_path(agent_id, task_sandbox_path)
+    else:
+        return []
+
+
+@retrying.retry(
+    wait_fixed=DEFAULT_RETRY_WAIT, stop_max_attempt_number=DEFAULT_RETRY_MAX_ATTEMPTS
+)
+def download_agent_path(
+    agent_id: str, agent_file_path: str, output_file_path: str, chunk_size: int = 8192
+) -> None:
+    stream = sdk_cmd.cluster_request(
+        "GET",
+        "/slave/{}/files/download?path={}".format(agent_id, agent_file_path),
+        stream=True,
+    )
+    with open(output_file_path, "wb") as f:
+        for chunk in stream.iter_content(chunk_size=chunk_size):
+            f.write(chunk)
+
+
+def download_sandbox_files(
+    agent_id: str,
+    sandbox: List[dict],
+    output_base_path: str,
+    patterns_to_download: List[str] = [],
+) -> List[dict]:
+    if not os.path.exists(output_base_path):
+        os.makedirs(output_base_path)
+
+    for task_file in sandbox:
+        task_file_basename = os.path.basename(task_file["path"])
+        for pattern in patterns_to_download:
+            if re.match(pattern, task_file_basename):
+                download_agent_path(
+                    agent_id,
+                    task_file["path"],
+                    os.path.join(output_base_path, task_file_basename),
+                )
+
+
+def download_task_files(
+    agent_id: str,
+    executor_sandbox_path: str,
+    task_id: str,
+    base_path: str,
+    patterns_to_download: List[str] = [],
+) -> List[dict]:
+    executor_sandbox = browse_executor_sandbox(agent_id, executor_sandbox_path)
+    task_sandbox = browse_task_sandbox(agent_id, executor_sandbox_path, task_id)
+
+    # Pod task: download both its logs and its parent executor's logs.
+    if task_sandbox:
+        output_task_directory = os.path.join(base_path, task_id, "task")
+        download_sandbox_files(
+            agent_id, executor_sandbox, output_task_directory, patterns_to_download
+        )
+
+        output_executor_directory = os.path.join(base_path, task_id, "executor")
```

Comment:
Does this mean we will duplicate the executor logs for tasks that share an executor? Would it not make sense to more closely map the folder structure on the cluster itself?

Reply:
> Does this mean we will duplicate the executor logs for tasks that share an executor?

Yes. This was intentional on my part. When I'm debugging a customer issue I'm generally looking at logs from a task-centric POV, so it made more sense to group them in this way, even though it is possible that some files are going to be duplicated. I'm going to add a TODO item for having the executor ID appended to the executor directory name, so that we have that information obvious just from looking at the file name.